cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -45,18 +45,49 @@ macro(get_apache_mirror)
   if(APACHE_MIRROR STREQUAL "")
     set(APACHE_MIRROR_INFO_URL "https://www.apache.org/dyn/closer.cgi?as_json=1")
     set(APACHE_MIRROR_INFO_PATH "${CMAKE_CURRENT_BINARY_DIR}/apache-mirror.json")
+
     if(EXISTS "${APACHE_MIRROR_INFO_PATH}")
       set(APACHE_MIRROR_DOWNLOAD_STATUS 0)
     else()
       file(DOWNLOAD "${APACHE_MIRROR_INFO_URL}" "${APACHE_MIRROR_INFO_PATH}"
            STATUS APACHE_MIRROR_DOWNLOAD_STATUS)
     endif()
+
     if(APACHE_MIRROR_DOWNLOAD_STATUS EQUAL 0)
       file(READ "${APACHE_MIRROR_INFO_PATH}" APACHE_MIRROR_INFO)
+
       string(REGEX MATCH "\"preferred\": \"[^\"]+" APACHE_MIRROR_PREFERRED
                    "${APACHE_MIRROR_INFO}")
       string(REGEX
-             REPLACE "\"preferred\": \"" "" APACHE_MIRROR "${APACHE_MIRROR_PREFERRED}")
+             REPLACE "\"preferred\": \"" "" APACHE_MIRROR_PREFERRED
+                     "${APACHE_MIRROR_PREFERRED}")
+
+      if(CMAKE_VERSION VERSION_LESS 3.7)
+        set(APACHE_MIRROR "${APACHE_MIRROR_PREFERRED}")
+      else()
+        # Append backup URLs to the preferred URL.
+        # Feature only available starting in 3.7
+        set(APACHE_MIRROR_LIMIT 5)
+
+        string(REGEX MATCH "\"http\": \\[.+\\]" APACHE_MIRROR_ALL "${APACHE_MIRROR_INFO}")
```

Comment:
This is clever, though I wonder (1) whether this is worth the trouble and (2) if it actually improves reliability. We still have a single point of failure on the mirror info URL. Why not just hard-code 5 mirrors?

Reply:
For one thing I've never seen an error due to the info URL, so anecdotally it seems much more reliable.

we can certainly hardcode five mirrors as a default case; the theory with the info URL is that it provides the nearest mirror to the requester which is ideal since we don't know where the package may be built relative to hard-coded mirrors

Reply:
I think that we can use all returned mirrors.
If the mirror info URL is failed, we use `https://apache.osuosl.org/` as the fallback mirror.
(It may be better that we can use `https://downloads.apache.org/` instead of `https://apache.osuosl.org/`.)